java/src/main/java/com/google/appengine/tools/pipeline/impl/PipelineManager.java

```diff
@@ -800,31 +759,67 @@ private static void runJob(RunJobTask task) {
     // HandleSlotFilledTasks for each of the slots that were immediately filled
     // by the running of the job.
     // See "http://goto/java-pipeline-model".
-    logger.finest("Job returned: " + returnValue);
-    registerSlotsWithBarrier(updateSpec, returnValue, rootJobKey, jobRecord.getKey(),
-        jobRecord.getQueueSettings(), currentRunGUID, finalizeBarrier);
+    log.finest("Job returned: " + returnValue);
+    registerSlotsWithBarrier(job.getUpdateSpec(), returnValue, rootJobKey, jobRecord.getKey(),
+        jobRecord.getQueueSettings(), job.getCurrentRunGUID(), finalizeBarrier);
     jobRecord.setState(State.WAITING_TO_FINALIZE);
-    jobRecord.setChildGraphGuid(currentRunGUID);
-    updateSpec.getFinalTransaction().includeJob(jobRecord);
-    updateSpec.getFinalTransaction().includeBarrier(finalizeBarrier);
+    jobRecord.setChildGraphGuid(job.getCurrentRunGUID());
+    job.getUpdateSpec().getFinalTransaction().includeJob(jobRecord);
+    job.getUpdateSpec().getFinalTransaction().includeBarrier(finalizeBarrier);
     backEnd.saveWithJobStateCheck(
-        updateSpec, jobRecord.getQueueSettings(), jobKey, State.WAITING_TO_RUN, State.RETRY);
+        job.getUpdateSpec(), jobRecord.getQueueSettings(), jobKey, State.WAITING_TO_RUN, State.RETRY);
+  }
+
+  private void inject(Job<?> job) {
+    if (job instanceof RootJobInstance) {
+      inject(((RootJobInstance) job).jobInstance);
+    }
+
+    Optional<Injectable> injectable = Optional.ofNullable(job.getClass().getAnnotation(Injectable.class));
+
+    if (injectable.isPresent()) {
+      DIUtil.inject(injectable.get().value().getName(), job);
+    }
+  }
+
+  /**
+   * deserializes the Job instance corresponding to the given JobRecord.
+   * @param jobRecord of execution of Job
+   * @return a Job instance, with runtime values set
+   */
+  Job<?> restore(JobRecord jobRecord) {
+    JobInstanceRecord jobExecutionRecord = jobRecord.getJobInstanceInflated();
+
+    if (jobExecutionRecord == null) {
+      throw new RuntimeException(
+        "Internal logic error:" + jobRecord + " does not have jobInstanceInflated.");
+    }
+
+    Job<?> job = jobExecutionRecord.getJobInstanceDeserialized();
+    invokePrivateJobMethod("setPipelineRunner", job, this);
+    UpdateSpec updateSpec = new UpdateSpec(jobRecord.getRootJobKey());
+    setJobRecord(job, jobRecord);
+    String currentRunGUID = GUIDGenerator.nextGUID();
+    setCurrentRunGuid(job, currentRunGUID);
+    setUpdateSpec(job, updateSpec);
```

Comment:
Yeah, maybe we could just drop all `invokeX` methods with reflection and just use setters:

```suggestion
    job.setPipelineRunner(this);
    UpdateSpec updateSpec = new UpdateSpec(jobRecord.getRootJobKey());
    job.setJobRecord(jobRecord);
    String currentRunGUID = GUIDGenerator.nextGUID();
    job.setCurrentRunGuid(currentRunGUID);
    job.setUpdateSpec(updateSpec);
```

Which is going to be way cheaper than current impl